pallets/circuit/src/optimistic/mod.rs

```diff
@@ -32,138 +69,88 @@ impl<T: Config> Optimistic<T> {
+
+        let optimistic_fsx_in_step = &crate::Pallet::<T>::get_current_step_fsx_by_security_lvl(
             local_ctx,
             SecurityLvl::Optimistic,
         );
+        // Slash loop
+        for fsx in optimistic_fsx_in_step {
+            // Look for invalid FSX cases to slash
+            if !fsx.is_successfully_confirmed() {
+                let sfx_bid = fsx.expect_sfx_bid();
+                let (insurance, reserved_bond) =
+                    (*sfx_bid.expect_insurance(), *sfx_bid.expect_reserved_bond());
 
-        let mut slashed_executors: Vec<(
-            &T::AccountId,
-            &InsuranceDeposit<T::AccountId, T::BlockNumber, EscrowedBalanceOf<T, T::Escrowed>>,
-        )> = vec![];
-        let mut repatriated_executors: Vec<&T::AccountId> = vec![];
-
-        for fsx in &optimistic_fsx_in_step {
-            let side_effect_id = fsx.input.generate_id::<SystemHashing<T>>();
-            if let Some((_id, insurance_request)) = local_ctx
-                .insurance_deposits
-                .iter()
-                .find(|(id, _)| *id == side_effect_id)
-            {
-                if let Some(bonded_relayer) = &insurance_request.bonded_relayer {
-                    if fsx.confirmed.is_some()
-                        && fsx
-                            .confirmed
-                            .as_ref()
-                            .expect("ensured exists in the same check")
-                            .err
-                            .is_none()
-                    {
-                        repatriated_executors.push(bonded_relayer)
-                    } else {
-                        slashed_executors.push((bonded_relayer, insurance_request))
-                    }
-                }
+                // First slash executor
+                <<T as Config>::Escrowed as EscrowTrait<T>>::Currency::slash_reserved(
+                    &sfx_bid.executor,
+                    insurance + reserved_bond,
+                );
+                slashed_reserve += reserved_bond;
+                slashed_counter += 1;
             }
         }
 
-        for (slash_executor, slash_insurance_request) in slashed_executors.iter() {
-            <<T as Config>::Escrowed as EscrowTrait<T>>::Currency::slash_reserved(
-                slash_executor,
-                slash_insurance_request.insurance,
-            );
-            <<T as Config>::Escrowed as EscrowTrait<T>>::Currency::deposit_creating(
-                &slash_insurance_request.requester,
-                slash_insurance_request.insurance,
-            );
-            <T as Config>::Executors::slash_bond(
-                slash_executor,
-                slash_insurance_request.reserved_bond,
-            );
+        let honest_counter = optimistic_fsx_in_step.len() - slashed_counter;
+        let repatriation_bonus_per_honest_fsx =
+            slashed_reserve / EscrowedBalanceOf::<T, T::Escrowed>::from(honest_counter as u32);
```

Comment:
I think we should rework this tbh. In the current implementation, some executors can end up not being reimbursed fully, which opens new ways to exploit this. All honest executors should receive the reward they bid on, all dishonest get slashed, and then we need to figure out what to do with any remaining funds